drivers/network/starfive/ethernet.c

```diff
@@ -0,0 +1,456 @@
+            update_ring_slot(&rx, rx.tail, (uint32_t) (buffer.io_or_offset & 0xffffffff),
+                            (uint32_t) (buffer.io_or_offset >> 32), 0, (uint32_t) (DESC_RXSTS_OWNBYDMA | BIT(24) | BIT(30)));
+            /* We will update the hardware register that stores the tail address. This tells
+            the device that we have new descriptors to use. */
+            THREAD_MEMORY_RELEASE();
+            *DMA_REG(DMA_CHAN_RX_TAIL_ADDR(0)) = rx_desc_base + sizeof(struct descriptor) * rx.tail;
+            rx.tail = (rx.tail + 1) % RX_COUNT;
+        }
+
+        net_request_signal_free(&rx_queue);
+        reprocess = false;
+
+
+        if (!net_queue_empty_free(&rx_queue) && !hw_ring_full(&rx, RX_COUNT)) {
+            net_cancel_signal_free(&rx_queue);
+            reprocess = true;
+        }
+    }
+}
+
+static void rx_return(void)
+{
+    bool packets_transferred = false;
+    while (!hw_ring_empty(&rx, RX_COUNT)) {
+        /* If buffer slot is still empty, we have processed all packets the device has filled */
+        volatile struct descriptor *d = &(rx.descr[rx.head]);
+        if (d->d3 & DESC_RXSTS_OWNBYDMA) {
+            break;
+        }
+        net_buff_desc_t buffer = rx.descr_mdata[rx.head];
+        THREAD_MEMORY_ACQUIRE();
+
+        if (d->d3 & DESC_RXSTS_ERROR) {
+            sddf_dprintf("ETH|ERROR: RX descriptor returned with error status %x\n", d->d3);
+
+            rx.descr_mdata[rx.tail] = buffer;
+            update_ring_slot(&rx, rx.tail, (uint32_t) (buffer.io_or_offset & 0xffffffff),
+                            (uint32_t) (buffer.io_or_offset >> 32), 0, (uint32_t) (DESC_RXSTS_OWNBYDMA | BIT(24) | BIT(30)));
+
+            /* We will update the hardware register that stores the tail address. This tells
+            the device that we have new descriptors to use. */
+            *DMA_REG(DMA_CHAN_RX_TAIL_ADDR(0)) = rx_desc_base + sizeof(struct descriptor) * rx.tail;
+            rx.head = (rx.head + 1) % RX_COUNT;
+        } else {
+            buffer.len = (d->d3 | 0xe);
+            int err = net_enqueue_active(&rx_queue, buffer);
+            assert(!err);
+            packets_transferred = true;
+            rx.head = (rx.head + 1) % RX_COUNT;
+        }
+    }
+
+    if (packets_transferred && net_require_signal_active(&rx_queue)) {
+        net_cancel_signal_active(&rx_queue);
+        microkit_notify(RX_CH);
+    }
+}
+
+static void tx_provide(void)
+{
+    bool reprocess = true;
+    int i = 0;
+    while (reprocess) {
+        while (!(hw_ring_full(&tx, TX_COUNT)) && !net_queue_empty_active(&tx_queue)) {
+            net_buff_desc_t buffer;
+            int err = net_dequeue_active(&tx_queue, &buffer);
+            assert(!err);
+
+            // For normal transmit descriptors, tdes2 needs to be set to generate an IRQ on transmit
+            // completion. We also need to provide the length of the buffer data in bits 13:0.
+            uint32_t tdes2 = DESC_TXCTRL_TXINT | buffer.len;
```

Comment:
You probably only want to interrupt when the next-to _last_ queued Tx buf is processed.  There's no point in interrupting for every frame.

Reply:
I agree, I'll try this change out. However this is the same method that is used in the meson driver, where we are setting the interrupt completion bit on every tx packet.